File: tests/cpp/test_doom_menu.cpp
// examples/doom/menu.md 검증 (C++17).
#include "doom_menu.hpp"
#include "minitest.hpp"

using namespace Doom;

TEST(menu_open_pushes_root) {
    MenuStack stack;
    auto root = std::make_shared<Menu>("Main", std::vector<MenuItem>{{"New Game", "Action", nullptr, nullptr}});
    stack.open(root);
    CHECK(stack.isOpen());
    CHECK(stack.current() == root.get());
}

TEST(menu_cancel_closes_when_stack_empties) {
    MenuStack stack;
    stack.open(std::make_shared<Menu>("Main", std::vector<MenuItem>{}));
    CHECK(stack.closeOneLevel());
    CHECK(!stack.isOpen());
}

TEST(menu_cancel_from_submenu_returns_to_parent) {
    MenuStack stack;
    auto parent = std::make_shared<Menu>("Main", std::vector<MenuItem>{});
    auto child = std::make_shared<Menu>("Options", std::vector<MenuItem>{});
    stack.open(parent);
    stack.stack.push_back(child);
    CHECK(!stack.closeOneLevel());
    CHECK(stack.current() == parent.get());
}

TEST(menu_selection_wraps_and_confirms) {
    bool actionRan = false;
    auto submenu = std::make_shared<Menu>("Options", std::vector<MenuItem>{{"Sound", "Action", nullptr, nullptr}});
    std::vector<MenuItem> items = {
        {"New Game", "Action", [&] { actionRan = true; }, nullptr},
        {"Options", "Submenu", nullptr, submenu},
        {"Fullscreen", "Toggle", nullptr, nullptr},
    };
    auto menu = std::make_shared<Menu>("Main", items);
    MenuStack stack;
    stack.open(menu);

    menu->selectedIndex = items.size() - 1;
    stack.moveSelection("Down");
    CHECK_EQ(menu->selectedIndex, 0u);

    stack.confirmSelection();
    CHECK(actionRan);

    menu->selectedIndex = 1;
    stack.confirmSelection();
    CHECK_EQ(stack.current()->title, std::string("Options"));

    // 새 스택에서 토글 확인.
    MenuStack stack2;
    auto menu2 = std::make_shared<Menu>("Main", items);
    stack2.open(menu2);
    menu2->selectedIndex = 2;
    auto* item = stack2.confirmSelection();
    CHECK(item->toggleValue);
}

TEST_MAIN()

File: tests/test_doom_menu.py
import unittest

from doom_menu import Menu, MenuItem, MenuStack


class MenuOpenCloseTest(unittest.TestCase):
    """menu.md의 '## Feature: 메뉴 열기/닫기' 검증."""

    def test_open_pushes_root_menu(self):
        stack = MenuStack()
        root = Menu("Main", [MenuItem("New Game", "Action")])
        stack.open(root)
        self.assertTrue(stack.isOpen)
        self.assertIs(stack.current, root)

    def test_cancel_pops_and_closes_when_stack_empties(self):
        stack = MenuStack()
        stack.open(Menu("Main", []))
        fullyClosed = stack.closeOneLevel()
        self.assertTrue(fullyClosed)
        self.assertFalse(stack.isOpen)

    def test_cancel_from_submenu_returns_to_parent(self):
        stack = MenuStack()
        parent = Menu("Main", [])
        child = Menu("Options", [])
        stack.open(parent)
        stack.stack.append(child)
        fullyClosed = stack.closeOneLevel()
        self.assertFalse(fullyClosed)
        self.assertIs(stack.current, parent)


class MenuSelectionTest(unittest.TestCase):
    """menu.md의 '## Feature: 메뉴 항목 선택' 검증."""

    def setUp(self):
        self.actions = []
        self.items = [
            MenuItem("New Game", "Action", action=lambda: self.actions.append("new")),
            MenuItem("Options", "Submenu", submenu=Menu("Options", [MenuItem("Sound", "Action")])),
            MenuItem("Fullscreen", "Toggle"),
        ]
        self.menu = Menu("Main", self.items)
        self.stack = MenuStack()
        self.stack.open(self.menu)

    def test_down_wraps_to_first_item(self):
        self.menu.selectedIndex = len(self.items) - 1
        self.stack.moveSelection("Down")
        self.assertEqual(self.menu.selectedIndex, 0)

    def test_up_wraps_to_last_item(self):
        self.menu.selectedIndex = 0
        self.stack.moveSelection("Up")
        self.assertEqual(self.menu.selectedIndex, len(self.items) - 1)

    def test_confirm_action_runs_it(self):
        self.menu.selectedIndex = 0
        self.stack.confirmSelection()
        self.assertEqual(self.actions, ["new"])

    def test_confirm_submenu_pushes_it(self):
        self.menu.selectedIndex = 1
        self.stack.confirmSelection()
        self.assertEqual(self.stack.current.title, "Options")

    def test_confirm_toggle_flips_value(self):
        self.menu.selectedIndex = 2
        item = self.stack.confirmSelection()
        self.assertTrue(item.toggleValue)
        self.stack.confirmSelection()
        self.assertFalse(item.toggleValue)


if __name__ == "__main__":
    unittest.main()
